spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py

```diff
@@ -302,81 +436,571 @@ def get_local_provenance_data(self):
+
+        # send initial attempt at sending all the data
+        self._send_all_data_based_packets(number_of_packets, data_to_write)
+
+        # verify completed
+        received_confirmation = False
+        time_out_count = 0
+        while not received_confirmation:
+            try:
+                # try to receive a confirmation of some sort from spinnaker
+                data = self._connection.receive(
+                    timeout=self.TIMEOUT_PER_RECEIVE_IN_SECONDS)
+                time_out_count = 0
 
+                # check which message type we have received
+                received_confirmation = self._outgoing_process_packet(
+                    data, data_to_write)
+
+            except SpinnmanTimeoutException:  # if time out, keep trying
+                # if the timeout has not occurred x times, keep trying
+                if time_out_count > TIMEOUT_RETRY_LIMIT:
+                    emergency_recover_state_from_failure(
+                        transceiver, self._app_id, self, self._placement)
+                    raise SpinnFrontEndException(
+                        TIMEOUT_MESSAGE.format(time_out_count))
+
+                # reopen the connection and try again
+                time_out_count += 1
+                remote_port = self._connection.remote_port
+                local_port = self._connection.local_port
+                local_ip = self._connection.local_ip_address
+                remote_ip = self._connection.remote_ip_address
+                self._connection.close()
+                self._connection = SCAMPConnection(
+                    local_port=local_port, remote_port=remote_port,
+                    local_host=local_ip, remote_host=remote_ip)
+
+                # if we have not received confirmation of finish, try to
+                # retransmit missing seq nums
+                if not received_confirmation:
+                    self._outgoing_retransmit_missing_seq_nums(data_to_write)
+
+    def _read_in_missing_seq_nums(self, data, data_to_write, position):
+        """ handles a missing seq num packet from spinnaker
+
+        :param data: the data to translate into missing seq nums
+        :param data_to_write: the data to write
+        :param position: the position in the data to write.
+        :rtype: None
+        """
+        # find how many elements are in this packet
+        n_elements = (len(data) - position) // WORD_SIZE
+
+        # store missing
+        self._missing_seq_nums_data_in[-1].extend(struct.unpack_from(
+            "<{}I".format(n_elements), data, position))
+
+        # determine if last element is end flag
+        if self._missing_seq_nums_data_in[-1][-1] == \
+                self.MISSING_SEQ_NUMS_END_FLAG:
+            del self._missing_seq_nums_data_in[-1][-1]
+            self._outgoing_retransmit_missing_seq_nums(data_to_write)
+        if (self._total_expected_missing_seq_packets == 0 and
```

Comment:
Just wondering if this ever gets activated given my mistrust of the count.  Additionally, if you have all the packets you probably also have the end flag... is there any way this wouldn't be true?

Reply:
I dunno.